In [1]:
#Kinetica webinar
#Kaggle Credit Scoring dataset
#based on Kinetica webcast, but couldn't implement API code exactly because don't have Anaconda Enterprise
#=> can't use from anacondda_enterprise import publish => can't use decorator publish
#so figured out my own way using Flask
#11/10/2017

import numpy as np
import pandas as pd
from sklearn.externals import joblib

#Flask already installed
from flask import Flask
from flask import jsonify
from flask import request

In [2]:
"""import json

REQUEST = json.dumps({
        'path': {},
        'args': {},
        'body': {}
    })
"""

"import json\n\nREQUEST = json.dumps({\n        'path': {},\n        'args': {},\n        'body': {}\n    })\n"

In [3]:
"""#validate the model works before passing it to web service
model = joblib.load('myModels/model.pkl')
guess = np.array([45, 0.802982129, 9120]).reshape(1,-1)
y_hat = model.predict(guess)
print y_hat
"""

"#validate the model works before passing it to web service\nmodel = joblib.load('myModels/model.pkl')\nguess = np.array([45, 0.802982129, 9120]).reshape(1,-1)\ny_hat = model.predict(guess)\nprint y_hat\n"

In [4]:
#creates an app object from Flask.
app = Flask(__name__)

In [5]:
#request defined
@app.route('/predict',methods=['POST'])

def predict(): 
    json_ = request.json
    query_df = pd.DataFrame(json_)
    
    ##prediction = clf.predict(query).tolist()
    ##return jsonify({'prediction': prediction})
    
    #guess = np.array([age_num, debt_num, income_num]).reshape(1,-1)
    y_hat = model.predict(query_df).tolist()
    
    return jsonify({'prediction': y_hat})

if __name__ == '__main__':
    model = joblib.load('myModels/model.pkl')
    print 'model loaded'
    app.run()

model loaded


To test this new function we can use the curl command 

single:  
curl -X POST http://localhost:5000/predict  -H "Content-Type: application/json" -d '[{"age": 17.0, "debtRatio": 0.8, "monthlyIncome": 9120}]'


multi:  
curl -X POST http://localhost:5000/predict  -H "Content-Type: application/json" -d '[{"age": 40.0, "debtRatio": 0.1218, "monthlyIncome": 2600},{"age": 45.0, "debtRatio": 0.80298, "monthlyIncome": 9120},{"age": 31.0, "debtRatio": 0.8247, "monthlyIncome": 3000}]'

### Incremental Web Service Coding

In [ ]:
"""
#for troubleshooting - verified can run web server
@app.route("/")
def hello():
    return "Hello API World!"

#app.run() starts the web server and ready to handle the request. 
#But at this moment, it can handle only one request.
if __name__ == "__main__":
    app.run()
"""

In [ ]:
"""
#for troubleshooting - verified can invoke method 'predict' without passing any variablea
@app.route('/predict',methods=['POST'])
def predict(): 
    age_num = 45
    debt_num = 0.802982129
    income_num = 9120
    guess = np.array([age_num, debt_num, income_num]).reshape(1,-1)
    y_hat = model.predict(guess).tolist()
    
    return jsonify({'prediction': y_hat})

if __name__ == '__main__':
    model = joblib.load('myModels/model.pkl')
    print 'model loaded'
    app.run()
"""

In [ ]:
"""
#kinetica's chunk - not fully implemented
def predict(age, debtRatio, monthlyIncome): 
    age_num = age[0]
    debt_num = debtRatio[0]
    income_num = monthlyIncome[0]
    guess = np.array[(age_num, debt_num, income_num)].reshape(1,-1)
    y_hat = model.predict(guess)
    
    return jsonify(y_hat)
"""